File: src/main/java/waterpower/common/block/ore/BlockMaterial.java
package waterpower.common.block.ore;

import java.awt.Color;

import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyEnum;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.color.IBlockColor;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import waterpower.Reference;
import waterpower.common.block.BlockWaterPower;
import waterpower.common.item.crafting.MaterialTypes;
import waterpower.common.recipe.IRecipeRegistrar;

public class BlockMaterial extends BlockWaterPower {
	private static final PropertyEnum<MaterialTypes> MATERIAL_TYPES = PropertyEnum.create("type", MaterialTypes.class);

    public BlockMaterial() {
        super("materialBlock", Material.ROCK, ItemBlockMaterial.class);

        registerOreDict();

        setDefaultState(blockState.getBaseState().withProperty(MATERIAL_TYPES, MaterialTypes.Zinc));
    }

    @Override
    protected BlockStateContainer createBlockState() {
    	return new BlockStateContainer(this, MATERIAL_TYPES);
    }
    
    @Override
    public IBlockState getStateFromMeta(int meta) {
    	return getDefaultState().withProperty(MATERIAL_TYPES, MaterialTypes.values()[meta]);
    }
    
    @Override
    public int getMetaFromState(IBlockState state) {
    	return state.getValue(MATERIAL_TYPES).ordinal();
    }

    @Override
    public int maxMetaData() {
        return MaterialTypes.values().length;
    }
    
    @Override
    public TileEntity createNewTileEntity(World worldIn, int meta) {
    	return null;
    }
    
    @Override
    public int colorMultiplier(IBlockState state, IBlockAccess worldIn, BlockPos pos, int tintIndex) {
    	int meta = getMetaFromState(state);
        MaterialTypes type = MaterialTypes.values()[meta];
    	return new Color(type.R, type.G, type.B, type.A).getRGB();
    }
    
    TextureAtlasSprite icon;

    public void registerIcons(TextureMap textureMap) {
    	icon = textureMap.registerSprite(new ResourceLocation(Reference.ModID + ":blocks/BLOCK"));
    }

    public void registerOreDict() {
        for (MaterialTypes value : MaterialTypes.values()) {
            IRecipeRegistrar.registerOreDict("block" + value.name(), new ItemStack(this, 1, value.ordinal()));
        }
    }
}

File: src/main/kotlin/waterpower/common/block/ore/BlockMaterial.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.ore

import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraftforge.fml.common.LoaderState
import net.minecraftforge.oredict.OreDictionary
import waterpower.annotations.Init
import waterpower.annotations.NewInstance
import waterpower.client.render.block.BlockColor
import waterpower.common.block.BlockEnum
import waterpower.common.init.WPBlocks
import waterpower.common.item.MaterialTypes
import java.awt.Color

@NewInstance(LoaderState.ModState.PREINITIALIZED)
class BlockMaterial : BlockEnum<MaterialTypes>("material_block", Material.ROCK, MaterialTypes::class.java, MaterialTypes.values()) {
    init {
        WPBlocks.material = this
        WPBlocks.blocks += this

        for (material in MaterialTypes.values()) {
            OreDictionary.registerOre("block" + material.name, getItemStack(material))
        }
    }

    override fun colorMultiplier(state: IBlockState, world: IBlockAccess?, pos: BlockPos?, index: Int): Int {
        val type = getTypeFromState(state)
        return Color(type.R, type.G, type.B, type.A).rgb
    }

    companion object {
        @JvmStatic
        @Init(LoaderState.ModState.INITIALIZED, side = 0)
        fun init() {
            Minecraft.getMinecraft().blockColors.registerBlockColorHandler(BlockColor, WPBlocks.material)
        }
    }

}